internal_ws/ykcompile/src/lib.rs

```diff
@@ -1185,7 +1192,8 @@ impl TraceCompiler {
     /// Compile a guard in the trace, emitting code to abort execution in case the guard fails.
     fn c_guard(&mut self, guard: &Guard, dl: DynamicLabel) {
         // FIXME some of the terminators from which we build these guards can have cleanup blocks.
-        // Currently we don't run any cleanup, but should we?
+        // Currently we don't run any cleanup, but should we? XXX does this mean we don't call
+        // drop? And, if so, what do we not call Drop on?
```

Comment:
The cleanup this is refering to is the unwind path of `-Cpanic=unwind` I think, but I am not sure.

Reply:
A lot of the terminators have "cleanup" blocks, which should be run in certain circumstances. This comment was written before we committed to `abort`, but more recently we've questioned that again...

Reply:
What are we gonna do about this comment then?

Reply:
I think cleanup blocks are not run if the abort strategy is used.

If so, we would only fix this if we moved to unwind strategy.

@bjorn3 may know.

Reply:
"run in certain circumstances" == "only if there's a panic"?

Reply:
Yes. I checked in the rustc code to be sure.

Looking at for example a MIR call terminator:
```
         /// Cleanups to be done if the call unwinds.                                                                                                                                                          
        cleanup: Option<BasicBlock>,                                                                                              
```

Reply:
Presumably it's mostly drops that happen on the cleanup path. I may be able to write a test which checks that no drop is run (which implies abort strategy) and if we change strategy later it would fail.

Thoughts?

Reply:
OK, perfect, that means that we can clarify this to something like "We assume the abort strategy so we don't call cleanup operations if the stack unwinds." I wonder if we can put that in as an `assert` (maybe even a static assert e.g. https://crates.io/crates/static_assertions ?) *iff* we're not compiled with abort? I wouldn't be surprised if that's visible as a `cfg` flag?

Reply:
There's a spanner in the works. As hard as I try, I cannot get tests to use the abort strategy. They appear to always use unwinding.

I had planned for a test like:
```
#[test]                                                                                                                                                                                                   
fn abort_strategy() {                                                                                                                                                                                     
    #[cfg(not(panic = "abort"))]                                                                                                                                                                          
    panic!("not using the abort strategy");                                                                                                                                                               
}                                    
```

This test always fails, regardless of strategy selection in `Cargo.toml` and I've verified this is only a problem with tests. Also verified outside of the Yorick build system. Tests always use unwind. I think the test harness relies on the ability to catch a panic to detect failures.

This may pose a problem for us, as it means our tests are in fact using unwind, but the JIT assumes abort...

I couldn't find much info on this, but maybe [this](https://github.com/rust-lang/rust/issues/67650) looks relevant. It looks like they abort strategy tests to work eventually, but it's not yet fully implemented.

Reply:
> I think the test harness relies on the ability to catch a panic to detect failures.

Oh, that is a good point! Hmm.

It looks like the only roadblock to `-Zpanic-abort-tests` is https://github.com/rust-lang/rust/issues/68936 ?

Alternatively, perhaps we do need to move to unwinding in the long term, in which case we could perhaps encode unwinding information that, for the time being, aborts execution (ironically...) so that we don't actually have to implement it now? I'm not really sure, and I don't have a firm opinion yet on this.

Reply:
Suppose that the interpreter (in a test) uses unwinding, and the JIT uses abort. Does it matter? It just means we can't catch panics in the JIT, doesn't it?

If this is still an issue, we should probably move to unwinding.

Reply:
I wonder if @bjorn3 has any thoughts?

Reply:
I'm not super familiar with the innards of libunwind (or whatever Rust uses for unwinding), but I assume that it just picks up metadata and does its thing. If so, I don't think you can different parts compiled differently, as such; but we can maybe set a flag that says "if you hit this call frame, abort". I'd be fine with that as a temporary hack.

Reply:
In normal rust [only one panic runtime is loaded at a time](https://www.ralfj.de/blog/2019/11/25/how-to-panic-in-rust.html). But our JIT is a separate entity with its own semantics.

A temporary hack is probably best, as unwinding out of jitcode sounds scary.

I'm going to need to investigate what a panic looks like in the JIT, so I don't think this should block closing this audit PR. I've raised an issue so it's not forgotten:
https://github.com/softdevteam/yk/issues/262

Reply:
I don't think rust-lang/rust#68936 is something to be really worried about. In cg_clif I always enable `-Zpanic-abort-tests`.